caddy/app.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package caddy provides an app and handler for Caddy Server (https://caddyserver.com/)
+// allowing it to turn any handler into one supporting the Vulcain protocol.
+
+package caddy
+
+import (
+	"errors"
+	"log/slog"
+
+	outline_prometheus "github.com/Jigsaw-Code/outline-ss-server/prometheus"
+	outline "github.com/Jigsaw-Code/outline-ss-server/service"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const outlineModuleName = "outline"
+
+func init() {
+	caddy.RegisterModule(ModuleRegistration{
+		ID:  outlineModuleName,
+		New: func() caddy.Module { return new(OutlineApp) },
+	})
+}
+
+type ShadowsocksConfig struct {
+	ReplayHistory int `json:"replay_history,omitempty"`
+}
+
+type OutlineApp struct {
+	ShadowsocksConfig *ShadowsocksConfig `json:"shadowsocks,omitempty"`
+
+	ReplayCache outline.ReplayCache
+	logger      *slog.Logger
+	Metrics     outline.ServiceMetrics
+	buildInfo   *prometheus.GaugeVec
+}
+
+var (
+	_ caddy.App         = (*OutlineApp)(nil)
+	_ caddy.Provisioner = (*OutlineApp)(nil)
+)
+
+func (OutlineApp) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{ID: outlineModuleName}
```

Comment:
Note we still need to provide a ModuleInfo with an accurate ID, so there is some duplication here now, but `New()` isn't used from it as far as I can tell. This is also why I'm not logging any errors here, since it is called and used in a few places.

Reply:
That's unfortunate. At least we don't have the two module objects co-existing anymore. Thanks for the changes.